File: telemetry/aitf/sdk/go/instrumentation/agent.go
package instrumentation

import (
	"context"
	"fmt"
	"sync"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/trace"

	"github.com/girdav01/AITF/sdk/go/semconv"
)

const agentTracerName = "instrumentation.agent"

// AgentInstrumentor traces AI agent operations.
type AgentInstrumentor struct {
	tracer trace.Tracer
}

// NewAgentInstrumentor creates a new agent instrumentor.
func NewAgentInstrumentor(tp trace.TracerProvider) *AgentInstrumentor {
	if tp == nil {
		tp = otel.GetTracerProvider()
	}
	return &AgentInstrumentor{tracer: tp.Tracer(agentTracerName)}
}

// SessionConfig configures an agent session span.
type SessionConfig struct {
	AgentName   string
	AgentID     string
	AgentType   string
	Framework   string
	Version     string
	Description string
	SessionID   string
	TeamName    string
}

// AgentSession manages spans within an agent session.
type AgentSession struct {
	span      trace.Span
	tracer    trace.Tracer
	agentName string
	sessionID string
	stepCount int
	mu        sync.Mutex
}

// TraceSession starts an agent session span.
func (a *AgentInstrumentor) TraceSession(ctx context.Context, cfg SessionConfig) (context.Context, *AgentSession) {
	if cfg.AgentType == "" {
		cfg.AgentType = semconv.AgentTypeAutonomous
	}
	if cfg.Framework == "" {
		cfg.Framework = semconv.AgentFrameworkCustom
	}

	attrs := []attribute.KeyValue{
		semconv.AgentNameKey.String(cfg.AgentName),
		semconv.AgentIDKey.String(cfg.AgentID),
		semconv.AgentTypeKey.String(cfg.AgentType),
		semconv.AgentFrameworkKey.String(cfg.Framework),
		semconv.AgentSessionIDKey.String(cfg.SessionID),
	}
	if cfg.Version != "" {
		attrs = append(attrs, semconv.AgentVersionKey.String(cfg.Version))
	}
	if cfg.Description != "" {
		attrs = append(attrs, semconv.AgentDescriptionKey.String(cfg.Description))
	}
	if cfg.TeamName != "" {
		attrs = append(attrs, semconv.AgentTeamNameKey.String(cfg.TeamName))
	}

	ctx, span := a.tracer.Start(ctx,
		fmt.Sprintf("agent.session %s", cfg.AgentName),
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)

	return ctx, &AgentSession{
		span:      span,
		tracer:    a.tracer,
		agentName: cfg.AgentName,
		sessionID: cfg.SessionID,
	}
}

// Step starts an agent step span as a child of the session.
func (s *AgentSession) Step(ctx context.Context, stepType string) (context.Context, *AgentStep) {
	s.mu.Lock()
	s.stepCount++
	currentStep := s.stepCount
	s.mu.Unlock()

	attrs := []attribute.KeyValue{
		semconv.AgentNameKey.String(s.agentName),
		semconv.AgentStepTypeKey.String(stepType),
		semconv.AgentStepIndexKey.Int(currentStep),
	}

	ctx, span := s.tracer.Start(ctx,
		fmt.Sprintf("agent.step.%s %s", stepType, s.agentName),
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)

	return ctx, &AgentStep{span: span}
}

// Delegate starts a delegation span.
func (s *AgentSession) Delegate(ctx context.Context, targetAgent, targetAgentID, reason, strategy string) (context.Context, trace.Span) {
	s.mu.Lock()
	s.stepCount++
	currentStep := s.stepCount
	s.mu.Unlock()

	attrs := []attribute.KeyValue{
		semconv.AgentNameKey.String(s.agentName),
		semconv.AgentStepTypeKey.String(semconv.AgentStepDelegation),
		semconv.AgentStepIndexKey.Int(currentStep),
		semconv.AgentDelegationTargetAgentKey.String(targetAgent),
		semconv.AgentDelegationTargetAgentIDKey.String(targetAgentID),
		semconv.AgentDelegationStrategyKey.String(strategy),
	}
	if reason != "" {
		attrs = append(attrs, semconv.AgentDelegationReasonKey.String(reason))
	}

	ctx, span := s.tracer.Start(ctx,
		fmt.Sprintf("agent.delegate %s -> %s", s.agentName, targetAgent),
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)
	return ctx, span
}

// End completes the session span.
func (s *AgentSession) End(err error) {
	s.mu.Lock()
	steps := s.stepCount
	s.mu.Unlock()

	s.span.SetAttributes(semconv.AgentSessionTurnCountKey.Int(steps))
	if err != nil {
		s.span.SetStatus(codes.Error, err.Error())
		s.span.RecordError(err)
	} else {
		s.span.SetStatus(codes.Ok, "")
	}
	s.span.End()
}

// Span returns the underlying OTel span.
func (s *AgentSession) Span() trace.Span { return s.span }

// AgentStep provides methods for setting step attributes.
type AgentStep struct {
	span trace.Span
}

// SetThought sets the agent's reasoning.
func (s *AgentStep) SetThought(thought string) {
	s.span.SetAttributes(semconv.AgentStepThoughtKey.String(thought))
}

// SetAction sets the planned action.
func (s *AgentStep) SetAction(action string) {
	s.span.SetAttributes(semconv.AgentStepActionKey.String(action))
}

// SetObservation sets the observation result.
func (s *AgentStep) SetObservation(observation string) {
	s.span.SetAttributes(semconv.AgentStepObservationKey.String(observation))
}

// End completes the step span.
func (s *AgentStep) End(err error) {
	if err != nil {
		s.span.SetStatus(codes.Error, err.Error())
		s.span.RecordError(err)
	} else {
		s.span.SetStatus(codes.Ok, "")
	}
	s.span.End()
}

// Span returns the underlying OTel span.
func (s *AgentStep) Span() trace.Span { return s.span }

File: telemetry/aitf/sdk/rust/src/instrumentation/agent.rs
//! Agent session instrumentation mirroring the Go `instrumentation/agent.go`.

use opentelemetry::global::{self, BoxedTracer};
use opentelemetry::trace::{Span, SpanKind, Status, Tracer};
use opentelemetry::KeyValue;

use crate::semconv::agent;

const TRACER_NAME: &str = "instrumentation.agent";
const DEFAULT_AGENT_TYPE: &str = "autonomous";
const DEFAULT_FRAMEWORK: &str = "custom";
const STEP_DELEGATION: &str = "delegation";

/// Configures an agent session span.
#[derive(Debug, Clone, Default)]
pub struct SessionConfig {
    /// Human-readable agent name.
    pub agent_name: String,
    /// Stable agent identifier.
    pub agent_id: String,
    /// Agent type (defaults to `autonomous`).
    pub agent_type: String,
    /// Framework (defaults to `custom`).
    pub framework: String,
    /// Optional agent version.
    pub version: String,
    /// Optional description.
    pub description: String,
    /// Session / conversation identifier.
    pub session_id: String,
    /// Optional team name.
    pub team_name: String,
}

impl SessionConfig {
    /// Convenience constructor from an agent name.
    pub fn new(agent_name: impl Into<String>) -> Self {
        Self {
            agent_name: agent_name.into(),
            ..Default::default()
        }
    }
}

/// Traces AI agent operations.
pub struct AgentInstrumentor<T = BoxedTracer> {
    tracer: T,
}

impl AgentInstrumentor<BoxedTracer> {
    /// Creates an instrumentor using the global tracer provider.
    pub fn new() -> Self {
        Self {
            tracer: global::tracer(TRACER_NAME),
        }
    }
}

impl Default for AgentInstrumentor<BoxedTracer> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T: Tracer> AgentInstrumentor<T>
where
    T: Clone,
{
    /// Creates an instrumentor from an explicit tracer.
    pub fn with_tracer(tracer: T) -> Self {
        Self { tracer }
    }

    /// Starts an agent session span. The caller must call
    /// [`AgentSession::end`] when done.
    pub fn trace_session(&self, mut cfg: SessionConfig) -> AgentSession<T> {
        if cfg.agent_type.is_empty() {
            cfg.agent_type = DEFAULT_AGENT_TYPE.to_string();
        }
        if cfg.framework.is_empty() {
            cfg.framework = DEFAULT_FRAMEWORK.to_string();
        }

        let mut attrs = vec![
            KeyValue::new(agent::NAME, cfg.agent_name.clone()),
            KeyValue::new(agent::ID, cfg.agent_id.clone()),
            KeyValue::new(agent::TYPE, cfg.agent_type.clone()),
            KeyValue::new(agent::FRAMEWORK, cfg.framework.clone()),
            KeyValue::new(agent::SESSION_ID, cfg.session_id.clone()),
        ];
        if !cfg.version.is_empty() {
            attrs.push(KeyValue::new(agent::VERSION, cfg.version.clone()));
        }
        if !cfg.description.is_empty() {
            attrs.push(KeyValue::new(agent::DESCRIPTION, cfg.description.clone()));
        }
        if !cfg.team_name.is_empty() {
            attrs.push(KeyValue::new(agent::TEAM_NAME, cfg.team_name.clone()));
        }

        let span = self
            .tracer
            .span_builder(format!("agent.session {}", cfg.agent_name))
            .with_kind(SpanKind::Internal)
            .with_attributes(attrs)
            .start(&self.tracer);

        AgentSession {
            span,
            tracer: self.tracer.clone(),
            agent_name: cfg.agent_name,
            step_count: 0,
        }
    }
}

/// Manages spans within an agent session.
pub struct AgentSession<T: Tracer> {
    span: T::Span,
    tracer: T,
    agent_name: String,
    step_count: i64,
}

impl<T: Tracer> AgentSession<T> {
    /// Starts an agent step span as a child of the session.
    pub fn step(&mut self, step_type: &str) -> AgentStep<T::Span> {
        self.step_count += 1;
        let attrs = vec![
            KeyValue::new(agent::NAME, self.agent_name.clone()),
            KeyValue::new(agent::STEP_TYPE, step_type.to_string()),
            KeyValue::new(agent::STEP_INDEX, self.step_count),
        ];
        let span = self
            .tracer
            .span_builder(format!("agent.step.{} {}", step_type, self.agent_name))
            .with_kind(SpanKind::Internal)
            .with_attributes(attrs)
            .start(&self.tracer);
        AgentStep { span }
    }

    /// Starts a delegation span.
    pub fn delegate(
        &mut self,
        target_agent: &str,
        target_agent_id: &str,
        reason: &str,
        strategy: &str,
    ) -> T::Span {
        self.step_count += 1;
        let mut attrs = vec![
            KeyValue::new(agent::NAME, self.agent_name.clone()),
            KeyValue::new(agent::STEP_TYPE, STEP_DELEGATION),
            KeyValue::new(agent::STEP_INDEX, self.step_count),
            KeyValue::new(agent::DELEGATION_TARGET_AGENT, target_agent.to_string()),
            KeyValue::new(
                agent::DELEGATION_TARGET_AGENT_ID,
                target_agent_id.to_string(),
            ),
            KeyValue::new(agent::DELEGATION_STRATEGY, strategy.to_string()),
        ];
        if !reason.is_empty() {
            attrs.push(KeyValue::new(agent::DELEGATION_REASON, reason.to_string()));
        }
        self.tracer
            .span_builder(format!(
                "agent.delegate {} -> {}",
                self.agent_name, target_agent
            ))
            .with_kind(SpanKind::Internal)
            .with_attributes(attrs)
            .start(&self.tracer)
    }

    /// Completes the session span, recording the turn count and status.
    pub fn end<E: std::fmt::Display>(mut self, result: Result<(), E>) {
        self.span
            .set_attribute(KeyValue::new(agent::SESSION_TURN_COUNT, self.step_count));
        match result {
            Ok(()) => self.span.set_status(Status::Ok),
            Err(e) => self.span.set_status(Status::error(e.to_string())),
        }
        self.span.end();
    }

    /// Returns a mutable reference to the underlying session span.
    pub fn span_mut(&mut self) -> &mut T::Span {
        &mut self.span
    }
}

/// Records attributes for a single agent step.
pub struct AgentStep<S: Span> {
    span: S,
}

impl<S: Span> AgentStep<S> {
    /// Sets the agent's reasoning.
    pub fn set_thought(&mut self, thought: impl Into<String>) {
        self.span
            .set_attribute(KeyValue::new(agent::STEP_THOUGHT, thought.into()));
    }

    /// Sets the planned action.
    pub fn set_action(&mut self, action: impl Into<String>) {
        self.span
            .set_attribute(KeyValue::new(agent::STEP_ACTION, action.into()));
    }

    /// Sets the observation result.
    pub fn set_observation(&mut self, observation: impl Into<String>) {
        self.span
            .set_attribute(KeyValue::new(agent::STEP_OBSERVATION, observation.into()));
    }

    /// Completes the step span.
    pub fn end<E: std::fmt::Display>(mut self, result: Result<(), E>) {
        match result {
            Ok(()) => self.span.set_status(Status::Ok),
            Err(e) => self.span.set_status(Status::error(e.to_string())),
        }
        self.span.end();
    }

    /// Returns a mutable reference to the underlying step span.
    pub fn span_mut(&mut self) -> &mut S {
        &mut self.span
    }
}
